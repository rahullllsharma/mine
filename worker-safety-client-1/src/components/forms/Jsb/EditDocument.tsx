import type { SelectPrimaryOption } from "@/components/shared/select/selectPrimary/SelectPrimary";
import type { ChangeEvent } from "react";
import type { FileProperties } from "./AttachmentsSection";
import { useState } from "react";
import { FileCategory } from "@/api/generated/types";
import ButtonPrimary from "@/components/shared/button/primary/ButtonPrimary";
import ButtonRegular from "@/components/shared/button/regular/ButtonRegular";
import Field from "@/components/shared/field/Field";
import FieldInput from "@/components/shared/field/fieldInput/FieldInput";
import FieldSelect from "@/components/shared/field/fieldSelect/FieldSelect";
import Paragraph from "@/components/shared/paragraph/Paragraph";

type SelectFieldOption = SelectPrimaryOption & {
  name: FileCategory;
};

// This is used only in this context
// Might be pushed to the backend in the future and fetched as needed
const documentCategories: SelectFieldOption[] = [
  { id: "1", name: FileCategory.Jha },
  { id: "2", name: FileCategory.Pssr },
  { id: "3", name: FileCategory.Hasp },
  { id: "4", name: FileCategory.Other },
];

export type UploadModalProps = {
  file: FileProperties;
  allowCategories?: boolean;
  onSave: (editedFile: EditedFile) => void;
  onCancel: () => void;
};

export type EditedFile = {
  id: string;
  displayName: string;
  name: string;
  category?: FileCategory;
};

export default function EditDocument({
  file,
  allowCategories = false,
  onSave,
  onCancel,
}: Readonly<UploadModalProps>): JSX.Element {
  const [editedName, setEditedName] = useState(file.name);
  const [documentCategory, setDocumentCategory] = useState<
    FileCategory | undefined
  >(file?.category);

  const saveDocumentHandler = () => {
    const editedFile: EditedFile = {
      id: file.id,
      displayName: editedName,
      name: file.name,
      category: documentCategory,
    };

    onSave(editedFile);
  };

  return (
    <>
      <FieldInput
        label="Name"
        placeholder="Add a name"
        defaultValue={editedName}
        onChange={(e: ChangeEvent<HTMLInputElement>) =>
          setEditedName(e.target.value)
        }
      />
      {allowCategories && (
        <FieldSelect
          label="Document type"
          className="mt-4"
          defaultOption={documentCategories.find(
            option => option.name === file.category
          )}
          options={documentCategories}
          onSelect={option => {
            const documentCategoryIndex = documentCategories.findIndex(
              doc => option.id === doc.id
            );

            setDocumentCategory(
              documentCategoryIndex > -1
                ? documentCategories[documentCategoryIndex]["name"]
                : undefined
            );
          }}
        />
      )}
      <Field label="File name" className="mt-4">
        <Paragraph text={file.name} />
      </Field>

      <Field label="Date / time uploaded" className="mt-4">
        <Paragraph text={`${file.date} ${file.time}`} />
      </Field>

      <div className="flex justify-end mt-10">
        <ButtonRegular className="mr-3" label="Cancel" onClick={onCancel} />
        <ButtonPrimary label="Save document" onClick={saveDocumentHandler} />
      </div>
    </>
  );
}
