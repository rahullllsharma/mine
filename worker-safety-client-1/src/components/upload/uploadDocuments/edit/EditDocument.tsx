import type { SelectPrimaryOption } from "@/components/shared/select/selectPrimary/SelectPrimary";
import type { ChangeEvent } from "react";
import type { UploadItem } from "../../Upload";
import { useState } from "react";
import ButtonPrimary from "@/components/shared/button/primary/ButtonPrimary";
import ButtonRegular from "@/components/shared/button/regular/ButtonRegular";
import Field from "@/components/shared/field/Field";
import FieldInput from "@/components/shared/field/fieldInput/FieldInput";
import FieldSelect from "@/components/shared/field/fieldSelect/FieldSelect";
import Paragraph from "@/components/shared/paragraph/Paragraph";

// This is used only in this context
// Might be pushed to the backend in the future and fetched as needed
const documentCategories: SelectPrimaryOption[] = [
  { id: "1", name: "JHA" },
  { id: "2", name: "PSSR" },
  { id: "3", name: "HASP" },
  { id: "4", name: "Other" },
];

export type UploadModalProps = {
  file: UploadItem;
  allowCategories?: boolean;
  onSave: (editedFile: EditedFile) => void;
  onCancel: () => void;
};

export type EditedFile = {
  id: string;
  displayName: string;
  name: string;
  category?: string;
};

export default function EditDocument({
  file,
  allowCategories = false,
  onSave,
  onCancel,
}: UploadModalProps): JSX.Element {
  const [editedName, setEditedName] = useState(file.name);
  const [documentCategory, setDocumentCategory] = useState<
    string | undefined | null
  >(file.category);

  const saveDocumentHandler = () => {
    let editedFile: EditedFile = {
      id: file.id,
      displayName: editedName,
      name: editedName,
    };

    if (documentCategory) {
      editedFile = { ...editedFile, category: documentCategory };
    }

    onSave(editedFile);
  };

  return (
    <>
      <FieldInput
        label="Name"
        placeholder="Add a name"
        defaultValue={editedName}
        onChange={(e: ChangeEvent<HTMLInputElement>) =>
          setEditedName(e.target.value)
        }
      />
      {allowCategories && (
        <FieldSelect
          label="Document type"
          className="mt-4"
          defaultOption={documentCategories.find(
            option => option.name === file.category
          )}
          options={documentCategories}
          onSelect={(option: SelectPrimaryOption) =>
            setDocumentCategory(option.name)
          }
        />
      )}
      <Field label="File name" className="mt-4">
        <Paragraph text={file.name} />
      </Field>
      <Field label="Date / time uploaded" className="mt-4">
        <Paragraph text={`${file.date} ${file.time}`} />
      </Field>

      <div className="flex justify-end mt-10">
        <ButtonRegular className="mr-3" label="Cancel" onClick={onCancel} />
        <ButtonPrimary label="Save document" onClick={saveDocumentHandler} />
      </div>
    </>
  );
}
