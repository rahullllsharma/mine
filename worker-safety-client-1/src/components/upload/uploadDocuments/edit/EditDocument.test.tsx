import { fireEvent, render, screen } from "@testing-library/react";
import { nanoid } from "nanoid";
import EditDocument from "./EditDocument";

const mockOnSave = jest.fn();
const mockOnCancel = jest.fn();

const file = {
  id: nanoid(),
  name: "CrewMember.xls",
  displayName: "Crew Details",
  size: "564 KB",
  date: "11/22/2021",
  time: "10:50PM",
  url: "path/to/file",
  signedUrl: "path/to/fileToDownload",
};

const fileWithCategory = { ...file, category: "JHA" };

describe(EditDocument.name, () => {
  describe("when it renders", () => {
    beforeEach(() => {
      render(
        <EditDocument file={file} onSave={mockOnSave} onCancel={mockOnCancel} />
      );
    });

    it("should match the snapshot", () => {
      const { asFragment } = render(
        <EditDocument file={file} onSave={mockOnSave} onCancel={mockOnCancel} />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    describe("when the user clicks the save button", () => {
      beforeEach(() => {
        jest.clearAllMocks();
      });

      it('should call the "onSave" callback with the same file name ', () => {
        fireEvent.click(screen.getByRole("button", { name: /save document/i }));
        expect(mockOnSave).toHaveBeenCalledWith({
          id: file.id,
          displayName: file.name,
          name: file.name,
        });
      });

      it("should create a document with the updated document name", () => {
        const inputElement = screen.getByRole("textbox");
        fireEvent.change(inputElement, { target: { value: "Test document" } });
        fireEvent.click(screen.getByRole("button", { name: /save document/i }));
        expect(mockOnSave).toHaveBeenCalledWith({
          id: file.id,
          displayName: "Test document",
          name: "Test document",
        });
      });
    });

    describe("when the user clicks the cancel button", () => {
      it('should call the "mockOnCancel" function', () => {
        fireEvent.click(screen.getByRole("button", { name: /cancel/i }));
        expect(mockOnSave).toHaveBeenCalled();
      });
    });
  });

  describe("when it renders with file type", () => {
    beforeEach(() => {
      render(
        <EditDocument
          file={fileWithCategory}
          onSave={mockOnSave}
          onCancel={mockOnCancel}
        />
      );
    });

    it("should match the snapshot", () => {
      const { asFragment } = render(
        <EditDocument
          file={fileWithCategory}
          onSave={mockOnSave}
          onCancel={mockOnCancel}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    describe("when the user clicks the save button", () => {
      beforeEach(() => {
        jest.clearAllMocks();
      });

      it('should call the "onSave" callback with the same file type', () => {
        fireEvent.click(screen.getByRole("button", { name: /save document/i }));
        expect(mockOnSave).toHaveBeenCalledWith({
          id: fileWithCategory.id,
          displayName: fileWithCategory.name,
          name: fileWithCategory.name,
          category: fileWithCategory.category,
        });
      });
    });
  });
});
