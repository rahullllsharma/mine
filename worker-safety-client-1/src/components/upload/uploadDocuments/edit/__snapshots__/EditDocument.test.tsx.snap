// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDocument when it renders should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
    >
      Name
    </label>
    <div
      class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"
    >
      <input
        aria-errormessage=""
        aria-invalid="false"
        class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 p-2"
        placeholder="Add a name"
        type="text"
        value="CrewMember.xls"
      />
    </div>
  </div>
  <div
    class="mt-4"
  >
    <label
      class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
    >
      File name
    </label>
    <p
      class="text-base text-neutral-shade-100"
    >
      CrewMember.xls
    </p>
  </div>
  <div
    class="mt-4"
  >
    <label
      class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
    >
      Date / time uploaded
    </label>
    <p
      class="text-base text-neutral-shade-100"
    >
      11/22/2021 10:50PM
    </p>
  </div>
  <div
    class="flex justify-end mt-10"
  >
    <button
      class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed text-base rounded-md font-semibold mr-3"
      type="button"
    >
      <div
        class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 px-2.5 py-2"
      >
        <span
          class="mx-1 truncate"
        >
          Cancel
        </span>
      </div>
    </button>
    <button
      class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed text-base rounded-md font-semibold text-white bg-brand-urbint-40"
      type="button"
    >
      <div
        class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 px-2.5 py-2"
      >
        <span
          class="mx-1 truncate"
        >
          Save document
        </span>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`EditDocument when it renders with file type should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
    >
      Name
    </label>
    <div
      class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"
    >
      <input
        aria-errormessage=""
        aria-invalid="false"
        class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 p-2"
        placeholder="Add a name"
        type="text"
        value="CrewMember.xls"
      />
    </div>
  </div>
  <div
    class="mt-4"
  >
    <label
      class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
    >
      File name
    </label>
    <p
      class="text-base text-neutral-shade-100"
    >
      CrewMember.xls
    </p>
  </div>
  <div
    class="mt-4"
  >
    <label
      class="block text-tiny md:text-sm text-neutral-shade-75 font-semibold mb-2 leading-normal"
    >
      Date / time uploaded
    </label>
    <p
      class="text-base text-neutral-shade-100"
    >
      11/22/2021 10:50PM
    </p>
  </div>
  <div
    class="flex justify-end mt-10"
  >
    <button
      class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed text-base rounded-md font-semibold mr-3"
      type="button"
    >
      <div
        class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 px-2.5 py-2"
      >
        <span
          class="mx-1 truncate"
        >
          Cancel
        </span>
      </div>
    </button>
    <button
      class="text-center truncate disabled:opacity-38 disabled:cursor-not-allowed text-base rounded-md font-semibold text-white bg-brand-urbint-40"
      type="button"
    >
      <div
        class="flex items-center justify-center h-full hover:bg-neutral-light-16 focus:bg-neutral-shade-7 active:bg-neutral-shade-7 px-2.5 py-2"
      >
        <span
          class="mx-1 truncate"
        >
          Save document
        </span>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
